### PR TITLE
Instrumentation: Log errors embedded within query data responses

### LINKED ISSUE #76140: Instrumentation: Query data errors embedded within query data response should be logged
Seems like more and more plugins are putting query data errors in the [data response](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go@v0.179.0/backend#DataResponse) instead of returning the error in the `QueryData` method. A plugin maybe doesn't log each of these errors and there's a chance that Grafana backend doesn't log this either. 

Suggest to extend the [plugin logger middleware](https://github.com/grafana/grafana/blob/main/pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go) to  log an error message for each error contained in each [data response](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go@v0.179.0/backend#DataResponse).

### PR DESCRIPTION
**Pull request by marefr**
**What is this feature?**

Logs errors embedded within query data responses.

**Why do we need this feature?**

Better understanding of errors embedded within query data responses

**Who is this feature for?**

Grafana operators

**Which issue(s) does this PR fix?**:

Fixes #76140 

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by marefr** [comment]
@bergquist do you have any quick thoughts/input on this one?

**Comment by xnyo** [comment]
This sounds reasonable to me. The only concern I have is about some plugins that may return sensitive information in the error, and that could then leak into the logs

**Comment by marefr** [comment]
> This sounds reasonable to me. The only concern I have is about some plugins that may return sensitive information in the error, and that could then leak into the logs

If that's the case we might have bigger problems since the information is returned to browser/client. The usageinsight feature already logs these as well

File: pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go [hunk 1]
```diff
@@ -76,6 +76,16 @@ func (m *LoggerMiddleware) QueryData(ctx context.Context, req *backend.QueryData
 		return innerErr
 	})
 
+	if err != nil {
+		return resp, err
+	}
+
+	for refID, dr := range resp.Responses {
+		if dr.Error != nil {
+			m.logger.Error("Query response error", "refID", refID, "error", dr.Error)
```
Comment (marefr) [inline pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go:85]: Should we perhaps somehow indicate this is coming from within query data response?

```suggestion
			m.logger.Error("Query data response with embedded error", "refID", refID, "error", dr.Error)
```

Comment (xnyo) [reply pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go:85]: I think `refID` log param is enough to indicate that the error is embedded in the response

**Review by xnyo — APPROVED** [review]
LGTM!

File: pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go (hunk 1 above)
Comment (marefr) [reply pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go:85]: Cool. I'm considering including status and status source here as well. Going to try it out first

Comment (marefr) [reply pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go]: I did change the message. WDYT?

File: pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go [hunk 2]
```diff
@@ -73,7 +73,19 @@ func (m *LoggerMiddleware) QueryData(ctx context.Context, req *backend.QueryData
...
-		return innerErr
+
+		if innerErr != nil {
+			return innerErr
+		}
+
+		ctxLogger := m.logger.FromContext(ctx)
+		for refID, dr := range resp.Responses {
+			if dr.Error != nil {
+				ctxLogger.Error("Partial data response error", "ref_id", refID, "error", dr.Error)
```
Comment (xnyo) [inline pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go:84]: Nit:

```suggestion
				ctxLogger.Error("Partial data response error", "refID", refID, "error", dr.Error)
```

I think camel case is preferred for log attributes.

https://github.com/grafana/grafana/blob/main/contribute/backend/instrumentation.md#naming-conventions

Not sure if `refID` vs `refId` tho

**Review by xnyo — COMMENTED** [review]
Awesome work, just one small comment

File: pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go (hunk 1 above)
Comment (xnyo) [reply pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go]: Looks good to me! :+1:

File: pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go (hunk 2 above)
Comment (marefr) [reply pkg/services/pluginsintegration/clientmiddleware/logger_middleware.go:84]: Good point thanks. Confused with the tracing attribute naming conventions :)

**Pull request merged**